doc/developer/design/20230224_persist_introspection.md

```diff
@@ -0,0 +1,111 @@
+
+If we then map each of them to a `RelationDesc`, we have everything we need to use them as inputs to dataflows in compute, making them usable in `SELECT`s, `INDEX`es, and `MATERIALIZED VIEW`s.
+
+This results in a timely operator that takes a `ShardId` and returns the following struct (TODO: or a single output of an `enum` type instead?):
+
+```rust
+pub struct PersistMetadata<G: Scope> {
+    // RelationDesc::empty()
+    //   .with_column("ts", ScalarType::UInt64.nullable(true))
+    pub since: Collection<G, Row, Diff>,
+    // RelationDesc::empty()
+    //   .with_column("lower", ScalarType::UInt64.nullable(true))
+    //   .with_column("upper", ScalarType::UInt64.nullable(true))
+    //   .with_column("since", ScalarType::UInt64.nullable(true))
+    //   .with_column("encoded_size_bytes", ScalarType::UInt64.nullable(false))
+    pub batches: Collection<G, Row, Diff>,
+    // RelationDesc::empty()
+    //   .with_column("id", ScalarType::String.nullable(false))
+    //   .with_column("since", ScalarType::UInt64.nullable(true))
+    //   .with_column("seqno", ScalarType::UInt64.nullable(false))
+    //   .with_column("last_heartbeat_timestamp_ms", ScalarType::UInt64.nullable(false))
+    //   .with_column("hostname", ScalarType::String.nullable(false))
+    //   .with_column("purpose", ScalarType::String.nullable(false))
+    pub leased_readers: Collection<G, Row, Diff>,
+    // RelationDesc::empty()
+    //   .with_column("id", ScalarType::String.nullable(false))
+    //   .with_column("since", ScalarType::UInt64.nullable(true))
+    //   .with_column("opaque", ScalarType::UInt64.nullable(true))
+    //   .with_column("hostname", ScalarType::String.nullable(false))
+    //   .with_column("purpose", ScalarType::String.nullable(false))
+    pub critical_readers: Collection<G, Row, Diff>,
+    // RelationDesc::empty()
+    //   .with_column("id", ScalarType::String.nullable(false))
+    //   .with_column("most_recent_write_upper", ScalarType::UInt64.nullable(false))
+    //   .with_column("last_heartbeat_timestamp_ms", ScalarType::UInt64.nullable(false))
+    //   .with_column("hostname", ScalarType::String.nullable(false))
+    //   .with_column("purpose", ScalarType::String.nullable(false))
+    pub writers: Collection<G, Row, Diff>,
+}
+```
+
+They are exposed in SQL as follows (strawman, requesting takes on this):
+- `SELECT * FROM PERSIST BATCHES METADATA FROM name_of_source_or_materialized_view`
+- This is only usable by the internal `mz_introspection`/`mz_system` users as well as by the system itself.
+- `BATCHES` can be one of several tokens, which each correspond to one of the persist metadata collections and a different `RelationDesc`.
+- Open question: This uses the "data" shard of the named item. Do we also want to expose the "remap" shard? What about other shards in the system?
```

Comment:
The remap shard is exposed as the data shard or the remap collection, so likely unnecessary

Reply:
👌 Good point! I left the "other shards" question as a catchall because I haven't done the audit of every shard in the system (e.g. what about the storage metrics ones) but if we have a catalog name for every shard, then this is a non-problem